File: services/dashboard/env_config/environment.py
"""
Environment-aware configuration module for dashboard service.
Detects whether running on Replit or production Ubuntu server.
"""

import os
from dataclasses import dataclass
from typing import Optional


@dataclass
class OpenAIConfig:
    """OpenAI API configuration"""
    api_key: str
    base_url: str
    model: str = "gpt-4o-mini"


@dataclass
class DatabaseConfig:
    """Database connection configuration"""
    url: str


def is_replit() -> bool:
    """Detect if running on Replit environment"""
    return (
        os.getenv("REPL_ID") is not None
        or os.getenv("REPLIT_CONNECTORS_HOSTNAME") is not None
    )


def get_openai_config() -> OpenAIConfig:
    """
    Get OpenAI configuration based on environment.
    
    - On Replit: Uses AI_INTEGRATIONS_* variables (Replit AI Integrations)
    - On Production: Uses OPENAI_API_KEY directly
    """
    if is_replit():
        # Replit AI Integrations (no API key needed)
        api_key = os.getenv("AI_INTEGRATIONS_OPENAI_API_KEY", "")
        base_url = os.getenv("AI_INTEGRATIONS_OPENAI_BASE_URL", "")
        
        if not api_key or not base_url:
            raise ValueError(
                "Running on Replit but AI_INTEGRATIONS_* env vars are missing. "
                "Please set up the OpenAI integration."
            )
        
        # Using gpt-4o for Replit environment
        model = os.getenv("AI_MODEL", "gpt-4o")
        
    else:
        # Production environment - use self-managed API key
        api_key = os.getenv("OPENAI_API_KEY", "")
        base_url = os.getenv("OPENAI_BASE_URL", "https://api.openai.com/v1")
        model = os.getenv("AI_MODEL", "gpt-4o-mini")
        
        if not api_key:
            raise ValueError(
                "OPENAI_API_KEY is required in production environment. "
                "Please add it to your .env file."
            )
    
    return OpenAIConfig(api_key=api_key, base_url=base_url, model=model)


def get_database_url(service_name: str) -> str:
    """
    Get database URL for a service.
    
    Args:
        service_name: Name of the service (e.g., 'jarvis', 'discord', 'streambot')
    
    Returns:
        Fully resolved database connection string
    """
    # Map service names to their database URLs
    url_mapping = {
        "jarvis": "JARVIS_DATABASE_URL",
        "discord": "DISCORD_DATABASE_URL",
        "streambot": "STREAMBOT_DATABASE_URL",
    }
    
    env_var = url_mapping.get(service_name.lower())
    if not env_var:
        raise ValueError(f"Unknown service: {service_name}")
    
    db_url = os.getenv(env_var)
    if not db_url:
        raise ValueError(
            f"{env_var} is not set. Please configure database connection."
        )
    
    # Verify it's not an unexpanded variable
    if "${" in db_url:
        raise ValueError(
            f"{env_var} contains unexpanded variable: {db_url}. "
            f"Please set the fully resolved connection string."
        )
    
    return db_url


def get_plex_config() -> dict:
    """Get Plex server configuration"""
    plex_url = os.getenv("PLEX_URL")
    plex_token = os.getenv("PLEX_TOKEN")
    
    if not plex_url or not plex_token:
        raise ValueError(
            "Plex configuration incomplete. "
            "PLEX_URL and PLEX_TOKEN are required."
        )
    
    return {
        "url": plex_url,
        "token": plex_token,
    }


# Environment info for debugging
def get_environment_info() -> dict:
    """Get current environment information for debugging"""
    return {
        "is_replit": is_replit(),
        "has_ai_integrations": bool(os.getenv("AI_INTEGRATIONS_OPENAI_BASE_URL")),
        "has_openai_key": bool(os.getenv("OPENAI_API_KEY")),
        "has_jarvis_db": bool(os.getenv("JARVIS_DATABASE_URL")),
        "has_plex_config": bool(os.getenv("PLEX_URL") and os.getenv("PLEX_TOKEN")),
    }

File: services/stream-bot/src/config/environment.ts
/**
 * Environment-aware configuration module for stream-bot service.
 * Detects whether running on Replit or production Ubuntu server.
 */

export interface OpenAIConfig {
  apiKey: string;
  baseURL: string;
  model: string;
}

export interface YouTubeConfig {
  clientId: string;
  clientSecret: string;
  redirectUri: string;
  useConnector: boolean;
}

export interface DatabaseConfig {
  url: string;
}

/**
 * Detect if running on Replit environment
 */
export function isReplit(): boolean {
  return !!(process.env.REPL_ID || process.env.REPLIT_CONNECTORS_HOSTNAME);
}

/**
 * Get OpenAI configuration based on environment.
 * 
 * - On Replit: Uses AI_INTEGRATIONS_* variables (Replit AI Integrations)
 * - On Production: Uses OPENAI_API_KEY directly
 */
export function getOpenAIConfig(): OpenAIConfig {
  if (isReplit()) {
    // Replit AI Integrations (no API key needed)
    const apiKey = process.env.AI_INTEGRATIONS_OPENAI_API_KEY || "";
    const baseURL = process.env.AI_INTEGRATIONS_OPENAI_BASE_URL || "";

    if (!apiKey || !baseURL) {
      throw new Error(
        "Running on Replit but AI_INTEGRATIONS_* env vars are missing. " +
          "Please set up the OpenAI integration."
      );
    }

    // Using gpt-4o for Replit environment (gpt-5 not available in production)
    const model = process.env.STREAMBOT_FACT_MODEL || "gpt-4o";

    return { apiKey, baseURL, model };
  } else {
    // Production environment - use self-managed API key
    const apiKey = process.env.OPENAI_API_KEY || process.env.STREAMBOT_OPENAI_API_KEY || "";
    const baseURL =
      process.env.OPENAI_BASE_URL ||
      process.env.STREAMBOT_OPENAI_BASE_URL ||
      "https://api.openai.com/v1";
    const model = process.env.STREAMBOT_FACT_MODEL || "gpt-4o";

    if (!apiKey) {
      throw new Error(
        "OPENAI_API_KEY or STREAMBOT_OPENAI_API_KEY is required in production environment. " +
          "Please add it to your .env file."
      );
    }

    return { apiKey, baseURL, model };
  }
}

/**
 * Auto-correct common OAuth redirect URI misconfigurations
 */
function correctRedirectUri(uri: string): string {
  let corrected = uri;
  
  // Fix: /api/auth/* -> /auth/* (wrong route prefix)
  if (corrected.includes('/api/auth/')) {
    corrected = corrected.replace('/api/auth/', '/auth/');
    console.log(`[OAuth Config] Auto-correcting redirect URI: removed /api prefix`);
  }
  
  // Fix: :5000 or :3000 port in HTTPS URL (should use 443 through Caddy)
  if (corrected.startsWith('https://') && /:(\d+)\//.test(corrected)) {
    corrected = corrected.replace(/:(\d+)\//, '/');
    console.log(`[OAuth Config] Auto-correcting redirect URI: removed port number from HTTPS URL`);
  }
  
  return corrected;
}

/**
 * Get YouTube OAuth configuration.
 * 
 * Both Replit and production use standard OAuth with client credentials.
 * YouTube Connector API (for developer access) is separate from user OAuth.
 */
export function getYouTubeConfig(): YouTubeConfig {
  const clientId = process.env.YOUTUBE_CLIENT_ID || "";
  const clientSecret = process.env.YOUTUBE_CLIENT_SECRET || "";
  // YOUTUBE_REDIRECT_URI is for platform connection (direct OAuth at /auth/youtube/callback)
  // YOUTUBE_SIGNIN_CALLBACK_URL is for passport signin (at /api/auth/youtube/callback)
  // For platform connection, default to /auth/youtube/callback
  const appUrl = process.env.APP_URL || "https://stream.rig-city.com";
  let redirectUri = process.env.YOUTUBE_REDIRECT_URI || `${appUrl}/auth/youtube/callback`;
  
  // Auto-correct common misconfigurations
  redirectUri = correctRedirectUri(redirectUri);

  if (!clientId || !clientSecret) {
    throw new Error(
      "YouTube OAuth not configured. YOUTUBE_CLIENT_ID and YOUTUBE_CLIENT_SECRET are required. " +
      "Get credentials from: https://console.cloud.google.com/apis/credentials"
    );
  }

  if (!redirectUri) {
    throw new Error(
      "YouTube OAuth redirect URI not configured. " +
      "YOUTUBE_REDIRECT_URI or YOUTUBE_SIGNIN_CALLBACK_URL is required."
    );
  }

  return {
    clientId,
    clientSecret,
    redirectUri,
    useConnector: false, // User OAuth always uses standard flow
  };
}

/**
 * Get database URL with validation
 */
export function getDatabaseURL(): string {
  const dbUrl = process.env.DATABASE_URL;

  if (!dbUrl) {
    throw new Error(
      "DATABASE_URL is not set. Please configure database connection."
    );
  }

  // Verify it's not an unexpanded variable
  if (dbUrl.includes("${")) {
    throw new Error(
      `DATABASE_URL contains unexpanded variable: ${dbUrl}. ` +
        `Please set the fully resolved connection string.`
    );
  }

  return dbUrl;
}

/**
 * Get environment information for debugging
 */
export function getEnvironmentInfo() {
  return {
    isReplit: isReplit(),
    hasAIIntegrations: !!process.env.AI_INTEGRATIONS_OPENAI_BASE_URL,
    hasOpenAIKey: !!(process.env.OPENAI_API_KEY || process.env.STREAMBOT_OPENAI_API_KEY),
    hasDatabase: !!process.env.DATABASE_URL,
    hasYouTubeConfig: !!(process.env.YOUTUBE_CLIENT_ID || isReplit()),
    nodeEnv: process.env.NODE_ENV || "development",
  };
}
